File: app/src/main/java/org/cerion/stockcharts/charts/views/ChartViewFactory.java
package org.cerion.stockcharts.charts.views;

import android.content.Context;
import android.graphics.Color;
import android.graphics.Paint;

import com.github.mikephil.charting.charts.BarLineChartBase;
import com.github.mikephil.charting.charts.Chart;
import com.github.mikephil.charting.charts.CombinedChart;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.components.AxisBase;
import com.github.mikephil.charting.components.Description;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.LegendEntry;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.CandleData;
import com.github.mikephil.charting.data.CandleDataSet;
import com.github.mikephil.charting.data.CandleEntry;
import com.github.mikephil.charting.data.CombinedData;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.formatter.IAxisValueFormatter;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;

import org.cerion.stocks.core.PriceList;
import org.cerion.stocks.core.charts.DataSet;
import org.cerion.stocks.core.charts.IDataSet;
import org.cerion.stocks.core.charts.IndicatorChart;
import org.cerion.stocks.core.charts.LineType;
import org.cerion.stocks.core.charts.PriceChart;
import org.cerion.stocks.core.charts.StockChart;
import org.cerion.stocks.core.charts.VolumeChart;
import org.cerion.stocks.core.model.Interval;

import java.math.BigDecimal;
import java.math.MathContext;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

class ChartViewFactory {

    private Context mContext;

    private static final int CHART_HEIGHT_PRICE = 800;
    private static final int CHART_HEIGHT = 400;
    private static DateFormat mDateFormat        = new SimpleDateFormat("MMM d, yy");
    private static DateFormat mDateFormatMonthly = new SimpleDateFormat("MMM ''yy");
    private Description mDesc = new Description();

    ChartViewFactory(Context context) {
        mContext = context;
        mDesc.setText("");
    }

    Chart getChart(StockChart chart, PriceList list) {
        if (chart instanceof PriceChart)
            return getPriceChart((PriceChart) chart, list);

        if (chart instanceof IndicatorChart)
            return getLineChart((IndicatorChart) chart, list);

        return getVolumeChart((VolumeChart) chart, list);
    }

    Chart getEmptyChart() {
        Chart chart = new LineChart(mContext);
        chart.setDescription(mDesc);
        chart.setMinimumHeight(ChartViewFactory.CHART_HEIGHT);
        return chart;
    }

    @SuppressWarnings("unchecked")
    private Chart getPriceChart(PriceChart pchart, PriceList list) {
        BarLineChartBase chart;
        List<IDataSet> sets = getDataSets(pchart, list);

        if(pchart.getCandleData() && pchart.canShowCandleData(list)) {
            chart = new CombinedChart(mContext);
            CombinedData data = new CombinedData();
            CandleData candleData = getCandleData(sets);
            data.setData(candleData);

            LineData lineData = getLineData(sets);
            data.setData(lineData);
            ((CombinedChart)chart).setDrawOrder(new CombinedChart.DrawOrder[]{CombinedChart.DrawOrder.CANDLE, CombinedChart.DrawOrder.LINE});
            chart.setData(data);
        } else {
            chart = new LineChart(mContext);
            LineData lineData = getLineData(sets);
            chart.setData(lineData);
        }

        setChartDefaults(chart, pchart, list);
        chart.setMinimumHeight(ChartViewFactory.CHART_HEIGHT_PRICE);

        if(pchart.getLogScale()) {
            YAxis axis = chart.getAxisRight();
            axis.setValueFormatter(getLogScaleYAxis());
        }

        setLegend(chart, sets);
        return chart;
    }

    private Chart getLineChart(IndicatorChart ichart, PriceList list) {
        LineChart chart = new LineChart(mContext);
        setChartDefaults(chart, ichart, list);
        chart.setMinimumHeight(ChartViewFactory.CHART_HEIGHT);

        List<IDataSet> sets = getDataSets(ichart, list);
        chart.setData(getLineData(sets));

        setLegend(chart, sets);
        return chart;
    }

    private Chart getVolumeChart(VolumeChart vchart, PriceList list) {
        CombinedChart chart = new CombinedChart(mContext);
        setChartDefaults(chart, vchart, list);

        if(vchart.getLogScale()) {
            YAxis axis = chart.getAxisRight();
            axis.setValueFormatter(getLogScaleYAxis());
        }

        List<IDataSet> dataSets = getDataSets(vchart, list);
        CombinedData data = new CombinedData();
        data.setData(getBarData(dataSets));
        data.setData(getLineData(dataSets));
        chart.setData(data);

        setLegend(chart, dataSets);
        return chart;
    }

    private void setChartDefaults(BarLineChartBase chart, StockChart stockchart, PriceList list) {
        chart.setDescription(mDesc);
        chart.setMinimumHeight(ChartViewFactory.CHART_HEIGHT);

        //Set Y axis
        chart.getAxisLeft().setDrawLabels(false);
        chart.getAxisRight().setPosition(YAxis.YAxisLabelPosition.INSIDE_CHART);
        chart.getAxisRight().setLabelCount(3, false);

        XAxis xaxis = chart.getXAxis();
        xaxis.setValueFormatter(getAxisFormatter(stockchart.getDates(list), list.getInterval()));
        xaxis.setPosition(XAxis.XAxisPosition.BOTTOM);

        // Always start at position 0 even if data set starts after that
        xaxis.setAxisMinimum(0);
    }

    private void setLegend(Chart chart, List<IDataSet> sets) {
        List<LegendEntry> entries = new ArrayList<>();

        String lastLabel = "";
        int lastColor = -1;
        for(IDataSet set : sets) {
            String label = set.getLabel();
            int color = set.getColor();
            LegendEntry entry = null;

            if(lastLabel.contentEquals(label)) {
                if(lastColor != color) {
                    entry = new LegendEntry(label, Legend.LegendForm.DEFAULT, Float.NaN, Float.NaN, null, color);
                    entries.get(entries.size() - 1).label = null; // label needs to go on the last one added
                }
            } else {
                entry = new LegendEntry(label, Legend.LegendForm.DEFAULT, Float.NaN, Float.NaN, null, color);
            }

            if(entry != null)
                entries.add(entry);

            lastLabel = label;
            lastColor = color;
        }

        Legend legend = chart.getLegend();
        legend.setCustom(entries);
        legend.setDrawInside(true);
        legend.setOrientation(Legend.LegendOrientation.VERTICAL);
        legend.setVerticalAlignment(Legend.LegendVerticalAlignment.TOP);
        legend.setWordWrapEnabled(false);
    }

    private List<IDataSet> getDataSets(StockChart chart, PriceList list) {
        chart.setPrimaryColors(Colors.CHART_PRIMARY);
        chart.setSecondaryColors(Colors.CHART_SECONDARY);
        return chart.getDataSets(list);
    }

    private BarData getBarData(List<IDataSet> sets) {
        List<IBarDataSet> result = new ArrayList<>();

        for(IDataSet curr : sets) {
            if(curr.getLineType() == LineType.BAR) {
                DataSet set = (DataSet)curr;
                ArrayList<BarEntry> entries = new ArrayList<>();
                for (int i = 0; i < set.getSize(); i++)
                    entries.add(new BarEntry(i, set.get(i)));

                BarDataSet dataSet = new BarDataSet(entries, set.getLabel());
                dataSet.setDrawValues(false);
                result.add(dataSet);
            }
        }

        return new BarData(result);
    }

    private LineData getLineData(List<IDataSet> sets) {
        List<ILineDataSet> result = new ArrayList<>();

        for(IDataSet curr : sets) {
            if(curr.getLineType() == LineType.LINE || curr.getLineType() == LineType.DOTTED) {
                DataSet set = (DataSet)curr;
                ArrayList<Entry> entries = new ArrayList<>();
                for (int i = 0; i < set.getSize(); i++) {
                    float point = set.get(i);
                    if(!Float.isNaN(point))
                        entries.add(new Entry(i, point));
                    //else if(i == 0)
                    //    entries.add(new Entry(0, 0));
                }

                LineDataSet lineDataSet = new LineDataSet(entries, set.getLabel());
                lineDataSet.setDrawCircles(false);
                lineDataSet.setDrawValues(false);
                lineDataSet.setColor(set.getColor());

                if(set.getLineType() == LineType.DOTTED) {
                    // https://github.com/PhilJay/MPAndroidChart/pull/2622
                    // This should be Transparent but using White because of bug
                    lineDataSet.setColor(Color.rgb(250,250,250));
                    lineDataSet.setDrawCircles(true);
                    lineDataSet.setCircleRadius(1f);
                    lineDataSet.setCircleColor(set.getColor());
                }

                result.add(lineDataSet);
            }
        }

        return new LineData(result);
    }

    private CandleData getCandleData(List<IDataSet> sets) {
        for(IDataSet set : sets) {
            if(set.getLineType() == LineType.CANDLE) {
                ArrayList<CandleEntry> entries = new ArrayList<>();
                org.cerion.stocks.core.charts.CandleDataSet cds = (org.cerion.stocks.core.charts.CandleDataSet)set;

                for (int i = 0; i < set.getSize(); i++) {
                    entries.add(new CandleEntry(i, cds.getHigh(i), cds.getLow(i), cds.getOpen(i), cds.getClose(i))); // order is high, low, open, close
                }

                CandleDataSet dataSet = new CandleDataSet(entries, set.getLabel());
                dataSet.setDrawValues(false);
                dataSet.setDecreasingColor(Colors.CANDLE_DOWN);
                dataSet.setDecreasingPaintStyle(Paint.Style.FILL);
                dataSet.setIncreasingColor(Colors.CANDLE_UP);
                dataSet.setIncreasingPaintStyle(Paint.Style.FILL);

                return new CandleData((dataSet));
            }
        }

        return null;
    }

    private IAxisValueFormatter getAxisFormatter(final Date[] dates, final Interval interval) {
        return new IAxisValueFormatter() {
            @Override
            public String getFormattedValue(float value, AxisBase axis) {
                int v = (int)value;
                if(interval == Interval.MONTHLY)
                    return mDateFormatMonthly.format(dates[v]);

                return mDateFormat.format(dates[v]);
            }
        };
    }

    private IAxisValueFormatter getLogScaleYAxis() {

        return new IAxisValueFormatter() {
            @Override
            public String getFormattedValue(float value, AxisBase axis) {
                // Round to 2 significant figures
                double actual = Math.exp(value);
                BigDecimal bd = new BigDecimal(actual);
                bd = bd.round(new MathContext(2));
                return bd.toPlainString();
            }
        };
    }
}

File: app/src/main/java/org/cerion/stockcharts/ui/charts/views/ChartViewFactory.kt
package org.cerion.stockcharts.ui.charts.views

import android.content.Context
import android.graphics.Color
import android.graphics.Paint
import com.github.mikephil.charting.charts.BarLineChartBase
import com.github.mikephil.charting.charts.Chart
import com.github.mikephil.charting.charts.CombinedChart
import com.github.mikephil.charting.charts.CombinedChart.DrawOrder
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.*
import com.github.mikephil.charting.data.*
import com.github.mikephil.charting.formatter.ValueFormatter
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet
import org.cerion.marketdata.core.charts.*
import org.cerion.marketdata.core.charts.CandleDataSet
import org.cerion.marketdata.core.charts.DataSet
import org.cerion.marketdata.core.model.Interval
import org.cerion.marketdata.core.model.OHLCVTable
import org.cerion.marketdata.core.platform.KMPDate
import org.cerion.stockcharts.R
import org.cerion.stockcharts.common.isDarkTheme
import java.math.BigDecimal
import java.math.MathContext
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import kotlin.math.exp

class ChartViewFactory(private val context: Context) {

    companion object {
        private const val CHART_HEIGHT_PRICE = 800
        private const val CHART_HEIGHT = 400
        private var dateFormat = DateTimeFormatter.ofPattern("MMM d, yy")
        private var dateFormatMonthly = DateTimeFormatter.ofPattern("MMM ''yy")
        private val blankDescription = Description().apply { text = "" }
    }

    private val _textColor =
            if (context.isDarkTheme())
                context.getColor(R.color.secondaryTextColor)
            else
                context.getColor(R.color.primaryTextColor)

    fun getChart(chart: StockChart, table: OHLCVTable): Chart<*> {
        return when(chart) {
            is PriceChart -> getPriceChart(chart, table)
            is IndicatorChart -> getLineChart(chart, table)
            is VolumeChart -> getVolumeChart(chart, table)
            else -> throw NotImplementedError()
        }
    }

    fun getEmptyChart(): Chart<*> {
        return LineChart(context).apply {
            description = blankDescription
            minimumHeight = CHART_HEIGHT
        }
    }

    private fun getPriceChart(pchart: PriceChart, table: OHLCVTable): Chart<*> {
        val chart: BarLineChartBase<*>
        val sets = getDataSets(pchart, table)
        if (pchart.candleData && pchart.canShowCandleData(table)) {
            val data = CombinedData()
            data.setData(getCandleData(sets))
            data.setData(getLineData(sets))

            chart = CombinedChart(context)
            setChartDefaults(chart, pchart, table)
            chart.drawOrder = arrayOf(DrawOrder.CANDLE, DrawOrder.LINE)
            chart.data = data
        }
        else {
            chart = LineChart(context)
            setChartDefaults(chart, pchart, table)
            val lineData = getLineData(sets)
            chart.data = lineData
        }

        chart.minimumHeight = CHART_HEIGHT_PRICE
        if (pchart.logScale)
            chart.axisRight.valueFormatter = logScaleYAxis

        setLegend(chart, sets)
        return chart
    }

    private fun getLineChart(ichart: IndicatorChart, table: OHLCVTable): Chart<*> {
        return LineChart(context).apply {
            setChartDefaults(this, ichart, table)
            minimumHeight = CHART_HEIGHT

            val sets = getDataSets(ichart, table)
            data = getLineData(sets)
            setLegend(this, sets)
        }
    }

    private fun getVolumeChart(vchart: VolumeChart, table: OHLCVTable): Chart<*> {
        return CombinedChart(context).apply {
            setChartDefaults(this, vchart, table)

            val dataSets = getDataSets(vchart, table)
            data = CombinedData().apply {
                setData(getBarData(dataSets))
                setData(getLineData(dataSets))
            }

            setLegend(this, dataSets)
            if (vchart.logScale)
                axisRight.valueFormatter = logScaleYAxis
        }
    }

    private fun setChartDefaults(chart: BarLineChartBase<*>, stockchart: StockChart, table: OHLCVTable) {
        if (chart.data != null)
            throw AssertionError("chart defaults should be set before data") // Needed for viewPortOffsets to work properly

        chart.apply {
            description = blankDescription
            minimumHeight = CHART_HEIGHT

            //Set Y axis
            axisLeft.setDrawLabels(false)
            axisRight.setPosition(YAxis.YAxisLabelPosition.INSIDE_CHART)
            axisRight.setLabelCount(3, false)
            axisRight.textColor = _textColor

            xAxis.valueFormatter = getAxisFormatter(stockchart.getDates(table), table.interval)
            xAxis.position = XAxis.XAxisPosition.BOTTOM_INSIDE
            // Always start at position 0 even if data set starts after that
            xAxis.axisMinimum = 0f
            xAxis.textColor = _textColor

            setViewPortOffsets(0f, viewPortHandler.offsetTop(), 0f, viewPortHandler.offsetBottom())
        }
    }

    private fun getDataSets(chart: StockChart, table: OHLCVTable): List<IDataSet> {
        return chart.getDataSets(table)
    }

    private fun getBarData(sets: List<IDataSet>): BarData {
        val result = mutableListOf<IBarDataSet>()

        sets.filter { it.lineType == LineType.BAR }
                .forEach {
                    val set = it as DataSet

                    val entries = set.mapIndexed { index, value ->
                        BarEntry(index.toFloat(), value)
                    }

                    val dataSet = BarDataSet(entries, set.label)
                    dataSet.setDrawValues(false)
                    dataSet.color = set.color
                    result.add(dataSet)
                }

        return BarData(result)
    }

    private fun getLineData(sets: List<IDataSet>): LineData {
        val result = mutableListOf<ILineDataSet>()

        for (curr in sets) {
            if (curr.lineType === LineType.LINE || curr.lineType === LineType.DOTTED) {
                val set = curr as DataSet
                val entries = set.mapIndexed { index, value ->
                    // TODO previous code checked for NaN, could be other issues if that comes up...
                    Entry(index.toFloat(), value)
                }

                val lineDataSet = LineDataSet(entries, set.label).apply {
                    setDrawCircles(false)
                    setDrawValues(false)
                    color = set.color

                    if (set.lineType === LineType.DOTTED) {
                        // https://github.com/PhilJay/MPAndroidChart/pull/2622 This should be Transparent but using White because of bug
                        color = Color.rgb(250, 250, 250)
                        setDrawCircles(true)
                        circleRadius = 1f
                        setCircleColor(set.color)
                    }
                }

                result.add(lineDataSet)
            }
        }

        return LineData(result)
    }

    private fun getCandleData(sets: List<IDataSet>): CandleData {

        for (set in sets) {
            if (set.lineType === LineType.CANDLE) {
                val entries = mutableListOf<CandleEntry>()
                val cds = set as CandleDataSet

                for (i in 0 until set.size)
                    entries.add(CandleEntry(i.toFloat(), cds.getHigh(i), cds.getLow(i), cds.getOpen(i), cds.getClose(i))) // order is high, low, open, close

                val dataSet = com.github.mikephil.charting.data.CandleDataSet(entries, set.label)
                dataSet.setDrawValues(false)
                dataSet.decreasingColor = context.getColor(R.color.negative_red)
                dataSet.decreasingPaintStyle = Paint.Style.FILL
                dataSet.increasingColor = context.getColor(R.color.positive_green)
                dataSet.increasingPaintStyle = Paint.Style.FILL

                return CandleData(dataSet)
            }
        }

        return CandleData()
    }

    private fun setLegend(chart: Chart<*>, sets: List<IDataSet>) {
        val entries = mutableListOf<LegendEntry>()
        var lastLabel = ""
        var lastColor = -1

        for (set in sets) {
            val label = set.label
            val color = set.color
            var entry: LegendEntry? = null
            if (lastLabel.contentEquals(label)) {
                if (lastColor != color) {
                    entry = LegendEntry(label, Legend.LegendForm.DEFAULT, Float.NaN, Float.NaN, null, color)
                    entries[entries.size - 1].label = null // label needs to go on the last one added
                }
            } else {
                entry = LegendEntry(label, Legend.LegendForm.DEFAULT, Float.NaN, Float.NaN, null, color)
            }

            if (entry != null)
                entries.add(entry)

            lastLabel = label
            lastColor = color
        }

        chart.legend.apply {
            setCustom(entries)
            setDrawInside(true)
            orientation = Legend.LegendOrientation.VERTICAL
            verticalAlignment = Legend.LegendVerticalAlignment.TOP
            isWordWrapEnabled = false
            textColor = _textColor
        }
    }

    private fun getAxisFormatter(dates: Array<KMPDate>, interval: Interval): ValueFormatter {
        return object : ValueFormatter() {
            override fun getFormattedValue(value: Float): String {
                val position = value.toInt()
                if (position >= dates.size)
                    return ""

                val date: LocalDate = dates[position].jvmDate
                return if (interval === Interval.MONTHLY)
                    dateFormatMonthly.format(date)
                else
                    dateFormat.format(date)
            }
        }
    }

    // Round to 2 significant figures
    private val logScaleYAxis: ValueFormatter = object : ValueFormatter() {
        override fun getFormattedValue(value: Float): String {
            // Round to 2 significant figures
            val actual = exp(value.toDouble())
            var bd = BigDecimal(actual)
            bd = bd.round(MathContext(2))
            return bd.toPlainString()
        }
    }
}